okhttp-tls/src/test/java/okhttp3/tls/HandshakeCertificatesTest.kt

```diff
@@ -0,0 +1,188 @@
+import java.util.concurrent.Future
+import javax.net.ServerSocketFactory
+import javax.net.SocketFactory
+import javax.net.ssl.SSLSocket
+import okhttp3.Handshake
+import okhttp3.Handshake.Companion.handshake
+import okhttp3.internal.closeQuietly
+import okhttp3.testing.PlatformRule
+import okio.ByteString.Companion.toByteString
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.RegisterExtension
+
+class HandshakeCertificatesTest {
+  @RegisterExtension
+  var platform = PlatformRule()
+
+  private lateinit var executorService: ExecutorService
+
+  private var serverSocket: ServerSocket? = null
+
+  @BeforeEach fun setUp() {
+    executorService = Executors.newCachedThreadPool()
+  }
+
+  @AfterEach fun tearDown() {
+    executorService.shutdown()
+    serverSocket?.closeQuietly()
+  }
+
+  @Test fun clientAndServer() {
+    platform.assumeNotConscrypt()
+    platform.assumeNotBouncyCastle()
+    val clientRoot = HeldCertificate.Builder()
+      .certificateAuthority(1)
+      .build()
+    val clientIntermediate = HeldCertificate.Builder()
+      .certificateAuthority(0)
+      .signedBy(clientRoot)
+      .build()
+    val clientCertificate = HeldCertificate.Builder()
+      .signedBy(clientIntermediate)
+      .build()
+    val serverRoot = HeldCertificate.Builder()
+      .certificateAuthority(1)
+      .build()
+    val serverIntermediate = HeldCertificate.Builder()
+      .certificateAuthority(0)
+      .signedBy(serverRoot)
+      .build()
+    val serverCertificate = HeldCertificate.Builder()
+      .signedBy(serverIntermediate)
+      .build()
+    val server = HandshakeCertificates.Builder()
+      .addTrustedCertificate(clientRoot.certificate)
+      .heldCertificate(serverCertificate, serverIntermediate.certificate)
+      .build()
+    val client = HandshakeCertificates.Builder()
+      .addTrustedCertificate(serverRoot.certificate)
+      .heldCertificate(clientCertificate, clientIntermediate.certificate)
+      .build()
+    val serverAddress = startTlsServer()
+    val serverHandshakeFuture = doServerHandshake(server)
+    val clientHandshakeFuture = doClientHandshake(client, serverAddress)
+    val serverHandshake = serverHandshakeFuture.get()
+    assertThat(listOf(clientCertificate.certificate, clientIntermediate.certificate))
+      .isEqualTo(serverHandshake.peerCertificates)
+    assertThat(listOf(serverCertificate.certificate, serverIntermediate.certificate))
+      .isEqualTo(serverHandshake.localCertificates)
+    val clientHandshake = clientHandshakeFuture.get()
+    assertThat(listOf(serverCertificate.certificate, serverIntermediate.certificate))
+      .isEqualTo(clientHandshake.peerCertificates)
+    assertThat(listOf(clientCertificate.certificate, clientIntermediate.certificate))
+      .isEqualTo(clientHandshake.localCertificates)
```

Comment:
Why not use `assertEquals` extension?
```kotlin
fun <@OnlyInputTypes T> assertEquals(expected: T, actual: T, message: String? = null) {
    asserter.assertEquals(message, expected, actual)
}
```

Reply:
I prefer fluent asserts with the actual first. 

Might need to migrate to kotlin.test asserts which aren't as nice to finish multiplatform, but that's out of scope here.

Reply:
Alright.